_posts/drafts/go-object-model.md

```diff
@@ -0,0 +1,569 @@
+type of the method receiver, on this case it is a (\*ObjectAdder).
+
+This is one of the more hard to understand parts of Go that I stumbled
+and I'm not going to be able to explain here right now, but it relates
+to [this](https://github.com/golang/go/wiki/MethodSets) and some other
+stuff that I assume are just implementation details.
+
+Anyway, going on, the result:
+
+```
+ObjectAdder.Add: func(*main.ObjectAdder, int, int) int
+ObjectAdder.Add: 1 + 1 = 2
+```
+
+As you can see, what Go actually does is to add a function on the type
+**\*ObjectAdder** that accepts a **\*ObjectAdder** as the first
+parameter. There is not method at all, it is just a function.
+
+What we see as an object in Go is actually a collection of functions associated
+to a type and syntactic sugar to pass the first argument for you.
+
+Which to be honest is like almost all object oriented languages
+is actually implemented.
+The good thing is that in Go this is 100% explicit, no magic, just some
+syntactic sugar. Go is really serious about being explicit and simple :-).
+
+This makes a lot of things more simple and uniform, the examples
+showed that. Passing a function or a method as a argument has no
+difference at all.
+
+Here is the full code of the final example:
+
+```go
+package main
+
+import "fmt"
+
+type Adder func(int, int) int
+
+// Same type as Adder
+func add(a int, b int) int {
+	return a + b
+}
+
+func abstractedAdd(a Adder, b int, c int) int {
+	return a(b, c)
+}
+
+type ObjectAdder struct{}
+
+func (o *ObjectAdder) Add(a int, b int) int {
+	return a + b
+}
+
+func main() {
+	var a Adder
+	fmt.Printf("Adder: %v\n", a)
+	a = add
+	fmt.Printf("Adder initialized: %v\n", a)
+	fmt.Printf("func: %d + %d = %d\n", 1, 1, abstractedAdd(a, 1, 1))
+	fmt.Printf("func: %d + %d = %d\n", 1, 1, abstractedAdd(add, 1, 1))
+
+	var o *ObjectAdder
+	fmt.Printf("func add: %T\n", add)
+	fmt.Printf("object.Add: %T\n", o.Add)
+	fmt.Printf("object: %d + %d = %d\n", 1, 1, abstractedAdd(o.Add, 1, 1))
+
+	fmt.Printf("ObjectAdder.Add: %T\n", (*ObjectAdder).Add)
+	fmt.Printf("ObjectAdder.Add: %d + %d = %d\n", 1, 1, (*ObjectAdder).Add(nil, 1, 1))
+}
+```
+
+This example is completely stateless. Another argument that I read on some
+discussions about Go is that functions should be used when there is no state
+involved, and objects and interfaces should be used when there is
+state involved.
+
+This question comes up specially when your Interface has just one method,
+why not a function ? I'm not very sure about using the state as an argument
+on this discussion, because as we have seen there is no difference between
+methods and functions, there is just functions, and functions can have
+state too.
+
+In a specific domain defining a guideline that all functions passed as
+parameters should be stateless may be useful, but it is dangerous to
+trust that in Go and it don't seem to be a general purpose guideline
+to choose between a function as parameter or a interface.
```

Comment:
I would write: "In a specific domain, dfining a guideline that all functions passed as parameters will be stateless may be useful, but it's dangerous to assume Go will respect that, and it doesn't seem to be a general purpose guideline to choose between a function as a parameter or an interface"

Does it make sense?